packages/dataviews/src/dataform.tsx

```diff
@@ -0,0 +1,83 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { TextControl } from '@wordpress/components';
+import { useEffect, useMemo } from '@wordpress/element';
+import { useDebouncedInput } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import type { Form, Field, NormalizedField } from './types';
+import { normalizeFields } from './normalize-fields';
+
+type DataFormProps< Item > = {
+	data: Item;
+	fields: Field< Item >[];
+	form: Form;
+	onUpdate: any; // TODO: fix this type.
+};
+
+type DataFormTextControlProps< Item > = {
+	data: Item;
+	field: NormalizedField< Item >;
+	onUpdate: any; // TODO: fix this type.
+};
+
+function DataFormTextControl< Item >( {
+	data,
+	field,
+	onUpdate,
+}: DataFormTextControlProps< Item > ) {
+	const { id, header, placeholder } = field;
+	const [ value, setValue, debouncedValue ] = useDebouncedInput(
+		field.getValue( { item: data } )
+	);
+
+	useEffect( () => {
+		onUpdate( ( prevItem: any ) => ( {
+			...prevItem,
+			[ id ]: debouncedValue,
+		} ) );
+	}, [ debouncedValue ] );
+
+	return (
+		<TextControl
+			label={ header }
+			placeholder={ placeholder }
+			value={ value }
+			onChange={ setValue }
+		/>
+	);
+}
+
+export default function DataForm< Item >( {
+	data,
+	fields,
+	form,
+	onUpdate,
+}: DataFormProps< Item > ) {
+	const visibleFields = useMemo(
+		() =>
+			normalizeFields(
+				fields.filter(
+					( { id } ) => !! form.visibleFields?.includes( id )
+				)
+			),
+		[ fields, form.visibleFields ]
+	);
+
+	return visibleFields.map( ( field ) => {
+		if ( field.type === 'text' ) {
```

Comment:
It would be cool to model a "registry" (even if closed to start with) of field types.

```
const text = {
   name: 'text',
   edit: DataFormTextControl
}
```

This can materialize in follow-ups as we start expanding.

Reply:
Pushed changes to use a registry-like structure at https://github.com/WordPress/gutenberg/pull/63032/commits/a29032179b200b28ef698b1a25989ff79117422b